mypy/semanal.py

```diff
@@ -545,7 +545,57 @@ def check_function_signature(self, fdef: FuncItem) -> None:
         elif len(sig.arg_types) > len(fdef.arguments):
             self.fail('Type signature has too many arguments', fdef, blocker=True)
 
+    def check_namedtuple_classdef(self, defn: ClassDef) -> Tuple[List[str], List[Type]]:
+        NAMEDTUP_CLASS_ERROR = 'Invalid statement in NamedTuple definition; ' \
+                               'expected "field_name: field_type"'
+        if self.options.python_version < (3, 6):
+            self.fail('NamedTuple class syntax is only supported in Python 3.6', defn)
+            return [], []
+        if len(defn.base_type_exprs) > 1:
+            self.fail('NamedTuple should be a single base', defn)
+        items = []  # type: List[str]
+        types = []  # type: List[Type]
+        for stmt in defn.defs.body:
+            if isinstance(stmt, AssignmentStmt):
+                if len(stmt.lvalues) > 1 or not isinstance(stmt.lvalues[0], NameExpr):
+                    self.fail(NAMEDTUP_CLASS_ERROR, stmt)
+                    continue
+                name = stmt.lvalues[0].name
+                if not isinstance(stmt.rvalue, TempNode):
+                    # x: int assigns rvalue to TempNode(AnyType())
+                    self.fail('NamedTuple does not support initial values', stmt)
+                if stmt.type is None:
+                    self.fail('NamedTuple field type must be specified', stmt)
+                    types.append(AnyType())
+                else:
+                    types.append(self.anal_type(stmt.type))
+                if name.startswith('_'):
+                    self.fail('NamedTuple field name cannot start with an underscore: {}'
+                              .format(name), stmt)
+                items.append(name)
+            else:
+                if (not isinstance(stmt, PassStmt) and
+                    not (isinstance(stmt, ExpressionStmt) and
+                         isinstance(stmt.expr, EllipsisExpr))):
+                    self.fail(NAMEDTUP_CLASS_ERROR, stmt)
+        return items, types
+
+    def analyze_namedtuple_classdef(self, defn: ClassDef) -> None:
+        node = self.lookup(defn.name, defn)
+        node.kind = GDEF  # TODO in process_namedtuple_definition also applies here
+        items, types = self.check_namedtuple_classdef(defn)
+        node.node = self.build_namedtuple_typeinfo(defn.name, items, types)
+
     def visit_class_def(self, defn: ClassDef) -> None:
+        # special case for NamedTuple
+        for base_expr in defn.base_type_exprs:
+            if isinstance(base_expr, NameExpr):
```

Comment:
Hm, this should probably also check for MemberExpr (similar to what's happening in check_namedtuple()).